Define the GitOps Principles (#48)
gitops-working-group/gitops-working-group: PRINCIPLES.md

<!-- see https://github.com/yzhang-gh/vscode-markdown/blob/master/README.md#table-of-contents -->
# GitOps Principles v0.1.0 <!-- omit in toc -->

## ⚠️ THIS DOCUMENT IS A WORK IN PROGRESS AND SUBJECT TO PUBLIC REVIEW BEFORE PUBLICATION <!-- omit in toc --> ⚠️ 

This document provides a concrete definition of GitOps principles.

> The GitOps Principles are vendor and implementation neutral, and aim to provide a common understanding of GitOps systems, to enable a common framework of understanding beyond individual opinion. A second aim is to encourage innovation by clarifying the technical outcomes rather than the code, tests, or organizational elements needed to achieve them. 

Reviewer: I struggle with `GitOps systems` as used in this statement - there are very meaningful components to GitOps that are not system rather process oriented. For instance if we consider the statement below: ```...a human or software agent will not interact with the running system and modify it directly. Instead, the agent will create a new declarative version of the desired state...``` states how the system must be interacted with.

> The GitOps Principles are vendor and implementation neutral, and aim to provide a common understanding of the specifications (qualities, characteristics, patterns?) that GitOps compliant systems and processes (workflows?) should satisfy.

Reviewer 2: Agree with @murillodigital about `GitOps systems`. 
Maybe simplify this even more by making shorter sentences and be less vague:

```suggestion
The GitOps Principles are vendor and implementation neutral. 
They provide common patterns for GitOps compliant systems and processes.
```

Reviewer 3: > They provide common patterns for GitOps compliant systems and processes. 

I am not sure that classifying some systems as GitOps compliant would be helpful in the long run. 

```suggestion
GitOps is a set of principles that governs how humans and technical systems interact in order to achieve the desired operational outcomes of repeatability, auditability, and visibility. 
```

Author: Have modified to integrate this discussion. Should hopefully resolve the lack of clarity.

2. [**The principle of immutable configuration versions**](#2-immutable-configuration-versions)

    _Desired State_ is stored in a way that supports versioning, immutability of versions, and retains a complete version history.
    We call systems that store desired state in this way _State Stores_.

3. [**The principle of continuous state reconciliation**](#3-continuous-state-reconciliation)

    Software agents continuously, and automatically, compare a system's _Actual State_ to its _Desired State_.
    If the actual and desired states differ, automated actions are immediately attempted to reconcile them.
    These differences could be due to the actual state drifting from the desired state, or the desired state changing intentionally.

4. [**The principle of operations through declaration**](#4-operations-through-declaration)

    When wishing to operate on a software system, a human or software agent will not interact with the running system and modify it directly.
    Instead, the agent will create a new declarative version of the desired state in the state store.

## Introduction

> The software systems that we manage vary widely; from battery powered embedded systems driven by microcontrollers, to globally distributed applications with millions of users. 
> Most systems are composed of subsystems themselves made of software.
> It is impossible to comprehensively outline all the specific practices for managing such a variety of systems.
> 
> However, despite the myriad differences, several important principles emerge that simplify the task of reliably managing and operating _all_ software systems at scale. GitOps attempts to capture some of these principles in a coherent framework.
> The GitOps principles can be applied to managing entire software systems or applied only to parts of larger systems.

Reviewer 3: ```suggestion
The systems that we manage vary widely; from battery-powered devices driven by microcontrollers to globally distributed systems with millions and sometimes billions of users. 

What is common across all systems is the need for human operators to make changes to the running state, it is during change that systems are most at risk of failure. 

GitOps is a framework in which all change is applied to a system in a consistent mechanism that both reduces risk and improves the situational awareness of human operators working inside the system.

GitOps can be applied to managing large systems or applied only to the smaller sub-system.
```

See also the [Rationale for GitOps](RATIONALE.md).

## Scope

GitOps concerns the verifiable behaviour of computer systems and their interfaces.
Specifically, GitOps is _not_ about human processes, and is not intended as a model for judging human organisational designs and operational practices.

The GitOps principles are to be used as guiding principles in the development of modern software and system operations. They do not form a concrete specification.

The GitOps principles are a direction, not a destination. They should be applied pragmatically. For example, whilst desirable to apply them strictly to an entire software systems, they can also be applied loosely to selected parts of larger systems as part of a progressive adoption.

> 

Reviewer 2: I do think that GitOps also makes sense in environments that don't have a healthy balance yet.
There, GitOps can act as an agent that reconciles the mismatch. ;)

> 

Author: I second this sentiment. (Good link about socio-technical systems too.) I wanted to make sure that the scope excluded human decision making explicitly. I've re-worded to reflect this and make it clearer. Do you think the meaning is clearer and more accurate now?

> 

Reviewer 4: I appreciated the original statement in line 83 and found it valuable/clarifying to directly callout the separation between our intent, the principles and a specification.   It also posed a small question with the inclusion of software development, whereas most GitOps discussion revolves around operations.  If I change the source code of a system to account for new business logic, does that fall under the umbrella of GitOps, which is heavily weighted on Ops?  How far left does do we envision GitOps spanning?

> 

Reviewer 5: @kmb385 currently this idea is captured in a few places in this PR (specifically "systems" or "subsystems"). Does this address your concern?

Here:

>By "Configuration", we mean _data that defines how a system or subsystem should behave_.
>
>For example, the same web server code may be running on thousands of different servers managed by hundreds of different companies.
>The behaviour of an individual webserver will differ based on how it is configured.
Configuration data is typically in the form of files or arguments to a computer program, but some systems may also currently use configuration databases or remote configuration services. Configuration also includes data about what version of code a software system should run, so software version information is also considered configuration.
>
>Together, the aggregate of all configuration data for a system form its "Desired State". The "Desired State" of a system is defined as data sufficient to recreate the system from nothing so that instances of the system are behaviorally indinstinguishable.

Also here:

>The definition of a _system_ can be quite broad, and may incorporate human as well as programmatic processes.
For example, is a company's sales process a system?
>Should GitOps be applied to it?
>Although a vision in which the GitOps practices are applied generally to all processes, human or otherwise is compelling, a more pragmatic approach is preferable, to avoid the risk of attempting to "boil the ocean".
>
>Instead, we should focus on subsystems where the Desired State is well defined, implement the GitOps principles there, and grow out to capture more systems from that initial subsystem. <!-- This progressive approach to GitOps is elaborated further in [the guide to adopting GitOps.md](ADOPTING_GITOPS.md). -->

The PR also retains this from the README:

>GitOps is fast becoming the methodology of choice for operating modern cloud native infrastructure and applications

> 

Reviewer 4: @scottrigby It does, I'm very new to the project and reading the definition of "software system" in the definitions section really helped.  My interpretation of that definition is that it does include the application source code.

## The GitOps Principles

### 1. Declarative configuration

<div style="background: #E3F2FD; border-left: 3px solid #0D47A1; padding:10px; margin-bottom:5px; font-style:italic">
A system managed by GitOps must have its <em>Desired State</em> expressed declaritively as data.

> This declarative data must be in a format writable and readable by both humans and software.

Author: I explicitly lay this out separately because I want to emphasise the fact that the definition is _data_, not code. I know "declarative" does this already, and saying "declarative data" is somewhat redundant, but it's not clear with everyone in practice and (Homoiconic languages aside), this is a big distinction.

> This declarative data must be in a format writable and readable by both humans and software.

Author: The language as is feels stilted though. Let me try to fix that.

</div>

#### What is a system's Desired State?

> _Configuration_ is a common feature of most software systems.

Reviewer 2: I find this change too.. nerdy. While true,  the original sentence "clicks" with most heads and does not sound too niche.

> _Configuration_ is a common feature of most software systems.

Reviewer 6: Technically this is not true. Many crypto systems are explicitly defined not to be configurable. I don’t equate hard coded with configuration, unless you include reflection or AOP or pointer hacking, etc.

By "Configuration", we mean _data that defines how a system or subsystem will behave and operate_.

> This configuration data is distinct and separate from the data a system will process.

Reviewer 3: ```suggestion
Configuration applies to the control surfaces of a system (i.e. control plane) 
```
I think the use of the work `data` is confusing - it has many different means depending on the context.

Reviewer: I agree with the removal of `data`, I would simply use `Configuration`, the use of `data` seems redundant and confusing.

I don't agree with the statement that configuration applies to control surfaces of a system only. Configuration may directly determine the behavior of the software itself (think configuration parameters that define runtime characteristics of an application -ie. heap size, feature flags, etc- which should also be declared as configuration).

To me the fundamental difference is _configuration defines **how** the system will behave_, as opposed to _data which defines **what** the system will process_

Reviewer 3: I think of control surfaces as the mechanism by which you apply the configuration to change behavior. 

How do we distinguish policy (give administrators admin access) from static state (list of users stored in git) and external state (pulling a list of users at reconcile stage) - Is it even necessary to do so?

Reviewer 3: > reproducible

I would avoid this term as it has a very specific meaning in some contexts e.g. binary reproducibility - I suggest repeatable

Author: Removing that sentence does not diminish this section and makes it less ambiguous. I amended to that effect to prevent misunderstanding. The clarification is not really needed in this context anyway.

For example, a Python program could verify a configuration file, while a C++ executable actually implements the declaration into a running system. Encoding the Desired State with a programming language ties the implementation to it, or forces other components to create a fully featured interpreter.

Secondly, verifying the correctness and self-consistency of data is significantly less complex than verifying the correctness of a program's behaviour, which is fundamentally undecidable. 
Verifying that a set of declarations is correct, however, _is_ decidable, even if sometimes computationally expensive.
As a general rule, this implies that the data language used to defined the Desired State should have no control flow structure, and consist exclusively of referentially transparent expressions without side effects. 
In other words, a human-readable data-serialization format.

It is also preferrable to use a widely supported language to define the Desired State for the sake of interoperability. For example, YAML, XML, JSON and TOML all have broad support and well defined specifications, although there are many other suitable candidates.

#### Why is human readability required?

> Operationally, it has been proven time and time again that the canonical Desired State of a system should be human-readable and writable.

Author: There are probably better references than the ones I added, but they should be a solid starting point.

This relative readability also implies that the users are relevant when evaluating whether a system follows this principle. For example, a Desired State defined with a rich grammar of S-Expressions would suite a team of developers with a background in functional programming, but would violate this principle if the majority of the team found the format incomprehensible.

Having a human-readable Desired State does not in any way preclude the use of rich tooling or graphical interfaces that facilitate Desired State generation and interpretation. It only require that the canonical source of truth be human readable and writable plain text.

#### How much of a system must be declared?

> Ideally, all of it; and the entire system can be recreated exclusively from its Desired State.

Reviewer: I think this is confusing in the context of application data. If I really wanted to recreate an entire system I'd have to consider recreating the application data that it requires to operate, and may even be interpreted in the context of disaster recovery (is GitOps a way to recover the latest known state of a system should disaster strike?) This also moves the conversation into scenarios that are perhaps beyond configuration, for instance, should DB migrations be declared as part of the desired state, and should I want to recreate a system matching that state must GitOps guarantee db schema consistency?

Reviewer 2: Definitely should mention that DB Migrations Question in a FAQ on the Website. Will be a commonly asked question.

Author: > is GitOps a way to recover the latest known state of a system should disaster strike?

The _latest configured behaviour_? Yes. The latest state including application data? Not at this time.

> Definitely should mention that DB Migrations Question in a FAQ on the Website. Will be a commonly asked question.

:+1: That's a super common discussion.

Author: > system matching that state must GitOps guarantee db schema consistency?

I think so, although this is genuinely difficult currently. Or rather, the db schema at a point in time should be declared. Moving to that schema from some arbitrary starting point is detail left to the reconciling agent. Currently, this will be in the form of a set of unidirectional migrations generated manually. That's an implementation detail though.

Author: I agree that we should mention application data here though.

Reviewer 7: @murillodigital wrote:
> should DB migrations be declared as part of the desired state?

Yes.

-----

@murillodigital wrote:
> must GitOps guarantee db schema consistency?

Should instead of must.

GitOps should guarantee database schema consistency.

-----

@bricef wrote:
> I think so, although this is genuinely difficult currently. Or rather, the db schema at a point in time should be declared. Moving to that schema from some arbitrary starting point is detail left to the reconciling agent. Currently, this will be in the form of a set of unidirectional migrations generated manually. That's an implementation detail though.

Agreed that database schema at a point in time should be declared.

Database migrations should be generated automatically, instead of manually.

Examples of implementation details:
1. [Ask HN: How does your development team handle database migrations?](https://news.ycombinator.com/item?id=19880334)
2. [Evolutionary Database Design](https://www.martinfowler.com/articles/evodb.html) by @pramodsadalage and @martinfowler at @thoughtworks using tools such as:
• https://github.com/liquibase/liquibase by @liquibase and @Datical
• https://github.com/flyway/flyway by @flyway and @red-gate
• https://github.com/mybatis/migrations by @mybatis
3. https://github.com/schemahero/schemahero by @schemahero, [donated by](https://www.replicated.com/blog/replicated-donates-schemahero-to-the-cncf/) @replicatedhq to @cncf
-----

From [GitOps Principles Committee - April 14, 2021 after 40:50](https://www.youtube.com/watch?v=zWxAZuHNGYM&t=2450s)

@bricef said:
> It's more about an aspirational direction, not necessarily what we do right now.

Agreed.

-----

@tonit wrote:
> DB Migrations Question in a FAQ

Agreed.

> That said, the definition of a "computer systems" can be quite broad and strays into human processes and systems as well. 

Reviewer: ```suggestion
The definition of _system_ can be quite broad, and may incorporate human as well as programmatic processes. Although a vision in which the GitOps practices are applied generally to all processes, human or otherwise, that participate in any one system, a more pragmatic approach is preferable, to avoid the risk of attempting to "boil the ocean".
```

For example, is a company's sales process a "computer system"? 
Likely in parts. 
Should GitOps be applied to the entire sales process including the human processes (without prejudice as to what those processes are)? 
Whilst such a vision is compelling, a more pragmatic approach is preferrable, otherwise we risk trying to boil the ocean.

Instead, we should focus on subsystems where the Desired State is well defined, implement the GitOps principles there, and grow out to capture more systems from that initial subsystem. <!-- This progressive approach to GitOps is elaborated further in [the guide to adopting GitOps.md](ADOPTING_GITOPS.md). -->

### 2. Immutable configuration versions

<div style="background: #E3F2FD; border-left: 3px solid #0D47A1; padding:10px; margin-bottom:5px; font-style:italic">
<em>Desired State</em> is stored in a way that supports versioning, immutability of versions, and retains a complete version history.
We call systems that store Desired State in this way <em>State Stores</em>.
</div>

#### What forms a version?

A version is the Desired State for a system as a whole. It is the canonical form of what we desire the system to be at a point in time.

> It is insufficient to version part of the Desired State or to version these parts in separate State Stores. Real software systems often have overarching behaviour that is the result of coupling between components. If the Desired State of these components were to change independently, it would be difficult to map a change in observed behaviour of our system to a single change in Desired State. Being able to make this 1:1 mapping is operationally benefitial, as we can then map behavioural issues of our system directly to the changes that occured. The utility of having the entire system defined in a single canonical location grows in proportion to the complexity and internal coupling of the system. A web of references to configuration data located in different locations is undesirable, as it makes understanding the desired state particularly difficult.

Reviewer: Not sure I follow this in the context of application configuration vs platform configuration as desired state. For instance, I may have some state in which I hold the declaration that I need some service running or crd available in which my application will depend (and maybe other applications as well), and some other state in which I specify the runtime parameters for one such specific application (say, endpoints or other runtime specific configuration for the application itself) - not sure its the best example but the point is, as part of a system there may be platform components as well as application components that will be interrelated, and following this statements it seems the argument is, anything that is somehow interrelated should be declared in the same state store. I think that such interpretation would be in the least not realistic.

Also I'd change `real software systems` to `production software systems` or `distributed systems`, can't think of a surreal or imaginary software system 😆

Author: 😄 Indeed.

Author: > the point is, as part of a system there may be platform components as well as application components that will be interrelated, and following this statements it seems the argument is, anything that is somehow interrelated should be declared in the same state store. I think that such interpretation would be in the least not realistic.

I agree that for all practical purposes, this is currently unrealistic, but the general point still stands. It is desirable to have all of a system's configuration in a single state store. I think changing "insufficient" to "undesirable" may be better. Do you think that's sufficient @murillodigital?

Reviewer: > 
> 
> > the point is, as part of a system there may be platform components as well as application components that will be interrelated, and following this statements it seems the argument is, anything that is somehow interrelated should be declared in the same state store. I think that such interpretation would be in the least not realistic.
> 
> I agree that for all practical purposes, this is currently unrealistic, but the general point still stands. It is desirable to have all of a system's configuration in a single state store. I think changing "insufficient" to "undesirable" may be better. Do you think that's sufficient @murillodigital?

How about changing the approach from that which is undesirable or suboptimal to that which is ideal:

`Ideally, the configuration for all interrelated components of a system should be stored in a single state store, in order to have a concise and consolidate view across all components that may have be subject to overarching behavior.`

Or something along those lines? @bricef

Reviewer 5: > How about changing the approach from that which is undesirable or suboptimal to that which is ideal

@murillodigital I like this approach flip

> Versions should be uniquely named. This need not be a semantically meanigful name. It is sufficient that each new version is attributed a name that identifies it uniquely. Once a new version has been created, it should be immutable. By this we mena that it should be impossible to modify the relationship between a version's unique name and its value of the Desired State. 

Reviewer: ```suggestion
Versions should be uniquely named. This need not be a semantically meaningful name. It is sufficient that each new version is attributed a name that identifies it uniquely. Once a new version has been created, it should be immutable. By this we mean that it should be impossible to modify the relationship between a version's unique name and its value of the Desired State. 
```

> All versions, except to very first,  should also have reference a predecessor or parent, which is another uniquely named version. This enables us to retain a history of the changes.

Reviewer: ```suggestion
All but the very first version should reference a predecessor or parent, which is another uniquely named version. This enables us to retain a history of the changes.
```

- changes as transactions

#### What are the responsibilities of a State store?
- In addition, state stores must preserve metadata information regarding the identity of agents that create new versions.
The Desired State of the new version, Who created it, when it was created and ideally, also _why_ it was created. - What Why, who when. 

- Even though we say git, it's actually only true when configured in a very particular way.

#### State stores as coordination points/ collab via common data interface
- Collab point. 
- Business rules and security rules about the acceptable state of the system shoudl be checked against any new proposed version of the desired configuration before such changes are accepted as a new version. (Gates before version)
- where human process is integrated
- the place where collaboration can be synchronised.
-->

### 3. Continuous state reconciliation

<div style="background: #E3F2FD; border-left: 3px solid #0D47A1; padding:10px; margin-bottom:5px; font-style:italic">
Software agents continuously, and automatically, compare a system's <em>Actual State</em> to its <em>Desired State</em>.

> If the actual and desired states differ, automated actions are immediately attempted to reconcile them.

Reviewer: Should a GitOps compliant system ALWAYS reconcile on identified drift or are other mechanisms possible to "close the loop" (alerting or otherwise?) I think this may be a barrier to a lot of organizations at some initial degrees of capability, although I agree this _should_ be the desired behavior. I guess an automated action could be an alert which triggers a human-lead attempt to reconciliation...

Reviewer 2: "Manually Closing the Loop". Definitely a FAQ point for the website.

Reviewer 3: I think desired state should also include the policy for when reconciliation is allowed to happen.

` I desire changes to Y to only occur during change windows, rollback on A, remediate on B and alert on C`
`I desire X to be Y`

These differences could be due to the actual state drifting from the desired state, or the desired state changing intentionally.
</div>

<!--

> - If the software agents fail to bring the system's state in line with its desired state, a human operator is notified.

Reviewer: Guess this is along the lines of my previous comment.

- Software agents continuously check that the running system under management matches the desired state configuration, and if it does not, immediately either take remedial action to bring the system back in line with stated expectations or, if this cannot be done, alert a human operator that the system is no longer meeting expectations
- Automated delivery: Delivery of the declarative descriptions, from the repository to runtime environment, is fully automated.
- Software Agents: Reconcilers maintain system state and apply the resources described in the declarative configuration.

> - Closed loop: Actions are performed on divergence between the version controlled declarative configuration and the actual state of the target system.

Reviewer 3: When I read closed loop I immediately think of https://en.wikipedia.org/wiki/Control_theory#Open-loop_and_closed-loop_(feedback)_control 

Most GitOps implementations are open reconciliation loops, the change is applied irrespective of the target state and the success or failure of the operation does not feedback into Git.  Some GitOps products e.g. Argo do try and build closed loops based but their feedback is internal only (either via UI or Kuebrentes events)  - Will open a new PR on this topic as I think it is an unsolved problem

Reviewer 5: It's true the _Current State_ doesn't feedback to the _Desired State_ store, but it does feed back to the _Software Agents_ so they know how the current and desired states do or do not diverge. This part of that wikipedia page you linked sounds like a correct description to me, where the "controller" map to GitOps _Software Agents_ :

> The definition of a closed loop control system according to the British Standard Institution is "a control system possessing monitoring feedback, the deviation signal formed as a result of this feedback being used to control the action of a final control element in such a way as to tend to reduce the deviation to zero."

WDYT?

- This automation should serve its human users.
  To this point, reconciliation automation for GitOps should include a way for humans to temporarily take back the reins as needed.
- Implies observability. State of the running system must be observable. Ideally in same format as the configuration so we can compare like to like. 
- >I agree. I think of this as an observability requirement and tend to say that "the current state of a running system must be observable" to mean the same thing. This is, in my opinion, required for state reconciliation. I didn't add a note there explicitly though, as that section was unfinished.
-->

### 4. Operations through declaration

<div style="background: #E3F2FD; border-left: 3px solid #0D47A1; padding:10px; margin-bottom:5px; font-style:italic">

> When wishing to operate on a software system, a human or software agent will not interact with the running system and modify it directly.

Reviewer: Isn't this incorrect? A software agent WILL interact with the system to apply the declaration that the human committed to the state store? (think an operator in a cluster)

Reviewer 2: ```suggestion
When wishing to operate on a software system, a human or computer program will not interact with the running system and modify it directly. Only the reconciliation-agent will operate on the software system.
```

Reviewer 3: ```suggestion
The mechanism in which change is applied to a system by either a human operator or another system is through a declarative change committed to a state store.  Reconciliation agents then apply these changes to the systems they control. 
```

Instead, the agent will create a new declarative version of the desired state in the state store.
</div>
<!--
- All normal operations should occur via the creation of a new uniquely named version, not through direct interaction ith the system under management
- Break glass - exceptions are acceptable and quite likely. and credentials to access and mutate the system must still be available in almost all cases.
-->

<!--
## General notes on the GitOps principles

- Tool and system agnosticicsm.
  
- It's ok to be pragmatic with implementations - We recognise that few real systems can be “100% GitOps”.
  Real world systems have a lifetime measured in years and must interact with many other systems created under alternative paradigms.
  As such they involve compromises and workarounds.
  
- lowest turtle isn't declarative, because the real world isn't.
  We aim for the lowest turtle to be as small as possible (imperative surface area )

> - Git as a satte store: - the "state store" can be a subfolder in a particular branch, and repo.

Reviewer: ```suggestion
- Git as a state store: - the "state store" can be a subfolder in a particular branch, and repo.
```

>   Principles only applies to the data used to configure a running system, not other data in the same git repository.

Reviewer: ```suggestion
  Principles only applies to the data used to configure a running system, not other data in the same git repository nor runtime application data.
```

Aligned with one of my comments above, GitOps is not about backup nor disaster recovery.

Author: Agreed. DR includes application data, GitOps doesn't.

    - [How much of a system must be declared?](#how-much-of-a-system-must-be-declared)
  - [2. Immutable configuration versions](#2-immutable-configuration-versions)
    - [What forms a version?](#what-forms-a-version)
  - [3. Continuous state reconciliation](#3-continuous-state-reconciliation)
  - [4. Operations through declaration](#4-operations-through-declaration)
- [See Also](#see-also)

## Summary

GitOps is a set of principles for operating and managing software systems.

When using GitOps, the desired state of a system or subsystem is defined declaratively as versioned, immutable data, and the running system's configuration is continuously derived from this data.

GitOps principles were derived from modern software operations but are rooted in pre-existing and widely adopted best practices. These principles are:

1. [**The principle of declarative configuration**](#1-declarative-configuration)

    A system managed by GitOps must have its _Desired State_ expressed declaritively as data in a format writable and readable by both humans and machines.

2. [**The principle of immutable configuration versions**](#2-immutable-configuration-versions)

>     _Desired State_ is stored in a way that supports versioning, immutability of versions, and retains a complete version history.

Reviewer 7: ```suggestion
    _Desired State_ is stored in a way that may support versioning and immutability of versions, and may retain a complete version history.
```
Suggested adding multiple "may"s because this principle shouldn't prevent practitioners from implementing Git, GitHub, BFG features such as:
1. [Configuring commit squashing for pull requests](https://docs.github.com/en/github/administering-a-repository/configuring-commit-squashing-for-pull-requests): `squash`
2. [Git Tools - Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History): `commit --amend`, `rebase`, `filter-branch`
3. [BFG is a simpler, faster alternative to git-filter-branch for cleansing bad data out of your Git repository history](https://rtyley.github.io/bfg-repo-cleaner/): `bfg`

**Why:**
• While versioning, immutability, completeness are recommended, practitioners should have the flexibility to opt-out.
• Seasoned `git` users are aware of situations that justify usages of `squash`, `commit --amend`, `rebase`, `filter-branch`, `bfg`.

This relates to https://github.com/gitops-working-group/gitops-working-group/discussions/93#discussioncomment-642197

Thank you @bricef @scottrigby @moshloop @o6uoq

Reviewer 3: I don't think any form of rewriting history should be allowed in the desired state store.

Rewriting history is perfectly acceptable *before* it merges into master (the desired state) but once it is desired state you can't erase it.

Likewise, filter-branch and bfg should not be allowed, if you commit a secret then you add a new commit to remove it and then revoke the secret. 

If you commit a large file or binary, then you either live with it, or you go off the GitOps path, rewrite history, and then go back on the GitOps path.

Reviewer 7: Can we agree to disagree?  Thank you @moshloop @roberthstrand @splushii

I believe rewriting history is appropriate to remove private information.

For example:
> **GitHub Private Information Removal Policy**
>
> We offer this private information removal process as an exceptional service only for high-risk content that violates [GitHub's Terms of Service](https://docs.github.com/en/github/site-policy/github-acceptable-use-policies#3-conduct-restrictions), such as when your security is at risk from exposed access credentials. This guide describes the information GitHub needs from you in order to process a request to remove private information from a repository.
>
> **[What is Private Information?](https://docs.github.com/en/github/site-policy/github-private-information-removal-policy#what-is-private-information)**
>
> For the purposes of this document, “private information” refers to content that (i) should have been kept confidential, _and_ (ii) whose public availability poses a specific or targeted security risk to you or your organization.
>
> "Security risk" refers to a situation involving exposure to physical danger, identity theft, or increased likelihood of unauthorized access to physical or network facilities.
>
> [Private information removal requests are appropriate for:](https://docs.github.com/en/github/site-policy/github-private-information-removal-policy#private-information-removal-requests-are-appropriate-for)
> - Access credentials, such as user names combined with passwords, access tokens, or other sensitive secrets that can grant access to your organization's server, network, or domain.
> - AWS tokens and other similar access credentials that grant access to a third party on your behalf. You must be able to show that the token does belong to you.
> - Documentation (such as network diagrams or architecture) that poses a specific security risk for an organization.
> • [Information](https://docs.github.com/en/github/site-policy/github-community-guidelines#doxxing-and-invasion-of-privacy) related to, and posing a security risk to, you as an individual (such as social security numbers or other government identification numbers).

4. [**The principle of operations through declaration**](#4-operations-through-declaration)

    The mechanism through which change is applied to a system by either a human operator or another system is through the creation of a new declarative version of the desired state in the state store, not through directly interacting with the running system.

## Introduction

The systems that we manage vary widely; from battery-powered devices driven by microcontrollers to globally distributed systems with millions and sometimes billions of users.

What is common across all systems is the need for human operators to make changes to the running state, it is during change that systems are most at risk of failure.

GitOps is a framework in which all change is applied to a system in a consistent mechanism that both reduces risk and improves the situational awareness of human operators working inside the system.

GitOps can be applied to managing large systems or applied only to the smaller sub-system.

See also the [Rationale for GitOps](RATIONALE.md).

## Scope

GitOps concerns the interaction between humans and technical systems, and between technical systems.

> GitOps is not concerned with processes of human decision making or organisation, only how human decisions about technical systems are recorded and applied.

Reviewer 3: I strongly disagree with this line. GitOps is all about human decision-making, i.e. a human deciding to make a change to a system and another human deciding that the change is of low enough risk to be merged. 

Part of the areas needing improvement is how human-decision making can be augmented by a better understanding of *how* a change could affect a system. (Tests / linting play a role here but usually disconnected from the system) i.e Will scaling down cause production workloads to be evicted or scaling up exceed my organizational budget

GitOps also has scaling problems in large organizations, and one of the goals of the WG should be to try and solve them e.g. 
* How do multiple teams in an organization interact on a single repo, 
* How is a system decomposed into multiple repositories 
* How are systematic changes across systems and repo's implemented and monitored
* How are changes promoted through environments while respecting change windows and organizational change policies. 

```suggestion
```

Reviewer 5: If I understand your comment correctly, this is not what was meant here. This is to clarify that the scope of GitOps is to define and reconcile technical systems, not human systems. Humans are always on some level in control of the automated processes. No [HAL](https://en.wikipedia.org/wiki/HAL_9000) or [Skynet](https://en.wikipedia.org/wiki/Skynet_(Terminator)) here LOL 😛  But seriously, we could probably word this better.

It is a structured process through which technical systems can be modified reliably.

> GitOps is _not_ intended as a model for judging human organisational designs.

Reviewer 3: Not sure how this could even be attempted using GItops?
```suggestion
```

Reviewer 5: For example, keeping an organizations employee reporting structure in git. But I'm not actually 100% sure we should list this as out of scope. For example, certain governmental and legal sources are kept in version control - where changes from those sources could be automatically reconciled to a public facing website using a GitOps software agent. Let's discuss this in the next co-working session If it remains controversial, we could leave it out of the first iteration.

Reviewer 6: CNCF and Kubernetes TOC being an example.

> The GitOps principles are to be used as guiding principles in the development of modern software and system operations. They do not form a concrete specification.
> 
> The GitOps principles are a _direction_, **not** a _destination_. They should be applied pragmatically. For example, whilst desirable to apply them strictly to an entire systems, they can also be applied selectively to sub-systems as part of a progressive adoption.
> 

Reviewer 3: ```suggestion
GitOps is used as a guiding principle for modern system operations. They are a _direction_, **not** a _destination_.

 As with all principles they should be applied pragmatically. For example, whilst desirable to apply them strictly to entire systems, they can also be applied selectively to sub-systems as part of a progressive adoption.
```

Reviewer 5: But GitOps can't be "used as a guiding principle", it's a set of principles. The intention of this doc is to define them.

Otherwise I like the rest of this suggestion 👍

For example, a Python program could verify a configuration file, while a C++ executable actually implements the declaration into a running system. Encoding the Desired State with a programming language ties the implementation to it, or forces other components to create a fully featured interpreter.

Secondly, verifying the correctness and self-consistency of data is significantly less complex than verifying the correctness of a program's behaviour, which is fundamentally undecidable. 
Verifying that a set of declarations is correct, however, _is_ decidable, even if sometimes computationally expensive.
As a general rule, this implies that the data language used to defined the Desired State should have no control flow structure, and consist exclusively of referentially transparent expressions without side effects. 
In other words, a human-readable data-serialization format.

It is also preferrable to use a widely supported language to define the Desired State for the sake of interoperability. For example, YAML, XML, JSON and TOML all have broad support and well defined specifications, although there are many other suitable candidates.

#### Why is human readability required?

> Operationally, it has been proven time and time again that the canonical Desired State of a system should be human-readable and writable. (See _The UNIX Philosophy (1995) by Mike Gancarz_ and _The Pragmatic Programmer (2000) Section 14 "The Power of Plain Text" by Andrew Hunt and David Thomas_)

Reviewer 3: Proven is a strong word, and I think we should wait until there is academic research that proves this 
```suggestion
The consensus among operators that the desired state of a system should be human-readable and writable is continuing to grow.
```

Reviewer 5: @moshloop Are you saying you find these resources to be not credible?

Is this sufficient?

```diff
- Operationally, it has been proven time and time again…
+ Operationally, consensus continues to grow that…
```

Reviewer 6: This is not true of ML systems. Human understanding or even observability is an area of ongoing research. Definitely not consensus yet. Growing, maybe. Desirable yes.
